### ISSUE TITLE
accessing/updating weights w/o optimizer?

### ISSUE THREAD

**Issue by zeroexcuses** [LaurentMazare/tch-rs#73]
This is not a big deal, more of an intellectual interest than an actual need.

1. I know that using optimizers (SGD, Adam, ...) we can easily update the weights of models. However, is there a way to access the model weights / update them manually ourselves ?

The example at https://pytorch.org/tutorials/beginner/examples_nn/two_layer_net_nn.html#sphx-glr-beginner-examples-nn-two-layer-net-nn-py uses the nn module to compute the gradients, but then tries to manually date it as follows:

```
model = torch.nn.Sequential(
    torch.nn.Linear(D_in, H),
    torch.nn.ReLU(),
    torch.nn.Linear(H, D_out),
)

while training :
    forward pass
    backward pass

    with torch.no_grad():
        for param in model.parameters():
            param -= learning_rate * param.grad

```

When I try to convert this over to tch-rs, the model becomes a tch::nn::Sequtnail, which also implements tch::Module.

Is there a way to either (1) grab the parameters or (2) update the parameters,
or does current design require that we use an optimizer and have the optimizer do the updating work?

[[ This seems this might be useful if the end user wants to write their own custom optimizer ]]

Thanks!

**Comment by LaurentMazare** [comment]
You can certainly do so. The [mnist-linear example](https://github.com/LaurentMazare/tch-rs/blob/master/examples/mnist/mnist_linear.rs) shows how to perform (2), but there is no real need for (1) in this case as the parameters are known.

(1) can also be automated to work on all the parameters in a var store, e.g.

```rust
fn my_module(p: nn::Path, dim: i64) -> impl nn::Module {
    let x1 = p.zeros("x1", &[dim]);
    let x2 = p.zeros("x2", &[dim]);
    nn::func(move |xs| xs * &x1 + xs.exp() * &x2)
}

#[test]
fn gradient_descent_test() {
    let vs = nn::VarStore::new(Device::Cpu);
    let my_module = my_module(vs.root(), 7);
    for _idx in 1..50 {
        let xs = Tensor::zeros(&[7], kind::FLOAT_CPU);
        let ys = Tensor::zeros(&[7], kind::FLOAT_CPU);
        let loss = (my_module.forward(&xs) - ys).pow(2).sum();
        for mut v in vs.trainable_variables() {
            v.zero_grad();
        }
        loss.backward();
        tch::no_grad(|| {
            for mut v in vs.trainable_variables() {
                v -= 0.1 * v.grad();
            }
        });
    }
}
```

Maybe it would be good to have a tutorial for such examples with a column for the python code and a column for the rust code.

**Comment by zeroexcuses** [comment]
Thanks!

I translate the full example as

```
use super::*;

fn main(opts: (tch::Kind, tch::Device)) {
    let (d_n, d_in, d_h, d_out) = (64, 1000, 100, 10);

    let x = tch::Tensor::randn(&[d_n, d_in], opts);
    let y = tch::Tensor::randn(&[d_n, d_out], opts);

    let vs = tch::nn::VarStore::new(opts.1);

    let model = tch::nn::seq()
        .add(tch::nn::linear(
            vs.root(),
            d_in,
            d_h,
            tch::nn::LinearConfig::default(),
        ))
        .add_fn(|xs| xs.relu())
        .add(tch::nn::linear(
            vs.root(),
            d_h,
            d_out,
            tch::nn::LinearConfig::default(),
        ));

    let learning_rate = 1e-4;

    for t in 0..500 {
        let y_pred = model.forward(&x);

        let loss = (y_pred - &y).pow(2).sum();

        for mut v in vs.trainable_variables() {
            v.zero_grad();
        }

        println!("{}: {:?}", t, loss);

        loss.backward();

        tch::no_grad(|| {
            for mut v in vs.trainable_variables() {
                v -= learning_rate * v.grad();
            }
        });
    }
}

```

I don't see how to use the mnist-linear technique as we define the model via `tch::nn::seq().add(...)`, but the mnist-linear technique seems to require to explicitly has the weights as tensors.

However, the `vs.trainable_variables()` solved all my problems. thanks!

=====

I have translated the examples at https://pytorch.org/tutorials/beginner/pytorch_with_examples.html , with the exception of the Tensorflow example and the "need support for custom backward" example.

See attached:
[pytorch_by_example_translation.zip](https://github.com/LaurentMazare/tch-rs/files/3407146/pytorch_by_example_translation.zip)

I'm not sure if a line by line conversion would have helped much. The most useful thing was being forced to read through parts of the tch-rs code (which I think is inevitable due to Rust/Python differences.)

**Comment by LaurentMazare** [comment]
Glad that it ended up working.
A line by line conversion may indeed not be super helpful as it would be a bit hard to read and most of the benefit comes from actually writing the code. So not sure how to help people with this learning curve, maybe having some kind of blog posts describing the experience of adapting some pytorch code to rust and the common hurdles along the way?

**Comment by zeroexcuses** [comment]
Let me translate a few more examples, logging points of frustration, and get back to you this.

Thanks again for the `vs.trainable_variables` insight!
